src/Cleanup.php

```diff
@@ -89,7 +89,32 @@ public function preMigration(string $mainRoleName): void
     {
         $sqls = [];
         $currentRole = $this->config->getTargetSnowflakeRole();
-        $this->destinationConnection->grantRoleToUser($this->config->getTargetSnowflakeUser(), $mainRoleName);
+
+        $mainRole = $this->destinationConnection->fetchAll(sprintf(
+            'SHOW ROLES LIKE %s',
+            QueryBuilder::quote($mainRoleName),
+        ));
+        $grantsToTargetUser = $this->destinationConnection->fetchAll(sprintf(
+            'SHOW GRANTS TO USER %s',
+            QueryBuilder::quoteIdentifier($this->config->getTargetSnowflakeUser()),
+        ));
+        $mainRoleExistsOnTargetUser = array_filter(
+            $grantsToTargetUser,
+            fn($v) => $v['role'] === $mainRoleName,
+        );
+
+        $hasMainRoleOwnership = array_filter(
+            $mainRole,
+            fn($v) => $v['owner'] === $this->config->getTargetSnowflakeUser()
+        );
+
+        if (!$mainRoleExistsOnTargetUser && !$hasMainRoleOwnership) {
+            throw new UserException('Main role is exists but not assign to migrate user.');
+        }
```

Comment:
Není špatně ta negace u `$mainRoleExistsOnTargetUser`?

```suggestion
        if ($mainRoleExistsOnTargetUser && !$hasMainRoleOwnership) {
            throw new UserException('Main role exists but is not assigned to migrate user.');
        }
```

Reply:
jo máš pravdu...ona ta celá podmínka by tu být nemusela, protože do toho stavu by se neměla appka dostat...tak jen pro jistotu